src/main/java/com/googleintern/wfm/ruleengine/model/RuleValidationReport.java

```diff
@@ -170,23 +205,70 @@ private static String convertSkillIdsToCsvString(List<Long> skillIds, List<Long>
-    wrongUserPoolAssignmentsBuilder.add(WRONG_USER_POOL_ASSIGNMENT_HEADER);
-    for (UserModel user : usersWithWrongAssignedPermissions) {
-      wrongUserPoolAssignmentsBuilder.add(
-          new String[] {
-            Long.toString(user.userId()),
-            Long.toString(user.workforceId()),
-            Long.toString(user.workgroupId()),
-            convertRoleIdsToCsvString(user.roleIds()),
-            convertSkillIdsToCsvString(user.skillIds(), user.roleSkillIds()),
-            convertPoolAssignmentsToCsvString(user.poolAssignments()),
-            convertPoolAssignmentsToCsvString(assignedPoolAssignmentsByUsers().get(user))
-          });
+  private ImmutableList<String[]> convertUsersWithLessAssignedPoolAssignmentsToCsvRows() {
+    return usersWithLessAssignedPermissions().stream()
+        .map(
+            user ->
+                new String[] {
+                  Long.toString(user.userId()),
+                  Long.toString(user.workforceId()),
+                  Long.toString(user.workgroupId()),
+                  convertRoleIdsToCsvString(user.roleIds()),
+                  convertSkillIdsToCsvString(user.skillIds(), user.roleSkillIds()),
+                  convertPoolAssignmentsToCsvString(
+                      Sets.difference(
+                              user.poolAssignments(), assignedPoolAssignmentsByUsers().get(user))
+                          .immutableCopy())
+                })
+        .collect(toImmutableList());
+  }
+
+  private ImmutableList<String[]> convertUsersWithMoreAssignedPoolAssignmentsToCsvRows() {
+    ImmutableSetMultimap<PoolAssignmentModel, RuleModel> rulesByPoolAssignments =
+        groupRulesByPoolAssignments();
+    return usersWithMoreAssignedPermissions().stream()
+        .map(
+            user ->
+                convertUserWithMoreAssignedPoolAssignmentsToCsvRow(
+                    user,
+                    rulesByPoolAssignments,
+                    Sets.difference(
+                            assignedPoolAssignmentsByUsers().get(user), user.poolAssignments())
+                        .immutableCopy()))
+        .collect(toImmutableList());
+  }
+
+  private String[] convertUserWithMoreAssignedPoolAssignmentsToCsvRow(
+      UserModel user,
+      ImmutableSetMultimap<PoolAssignmentModel, RuleModel> rulesByPoolAssignments,
+      ImmutableSet<PoolAssignmentModel> wrongAssignedPoolPermissions) {
+    return new String[] {
+      Long.toString(user.userId()),
+      Long.toString(user.workforceId()),
+      Long.toString(user.workgroupId()),
+      convertRoleIdsToCsvString(user.roleIds()),
+      convertSkillIdsToCsvString(user.skillIds(), user.roleSkillIds()),
+      convertPoolAssignmentsToCsvString(wrongAssignedPoolPermissions),
+      convertRulesToCsvString(
+          findRulesAssignedMorePermissions(
+              rulesByPoolAssignments, user, wrongAssignedPoolPermissions))
+    };
+  }
```

Comment:
Can these be combined into two methods that take parameters? There is quite a bit of repeated code here, the main content for the row is the same, with the difference being the last column:

convertUsersWithIncorrectPoolAssignmentsToCsvRows(bool isMorePermissions)
and same for a single user.